go/plugins/internal/gemini/gemini.go

```diff
@@ -26,3 +40,475 @@ var (
+			if err != nil {
+				return nil, err
+			}
+			content = append(content, &genai.Content{
+				Parts: parts,
+			})
+		}
+
+		r, err := genai.Models.EmbedContent(*client.Models, ctx, name, content, embedConfig)
+		if err != nil {
+			return nil, err
+		}
+		var res ai.EmbedResponse
+		for _, emb := range r.Embeddings {
+			res.Embeddings = append(res.Embeddings, &ai.DocumentEmbedding{Embedding: emb.Values})
+		}
+		return &res, nil
+	})
+}
+
+// Generate requests a generate call to the specified model with the provided
+// configuration
+func Generate(
+	ctx context.Context,
+	client *genai.Client,
+	model string,
+	input *ai.ModelRequest,
+	cb func(context.Context, *ai.ModelResponseChunk) error,
+) (*ai.ModelResponse, error) {
+	gc, err := convertRequest(client, model, input)
+	if err != nil {
+		return nil, err
+	}
+
+	var contents []*genai.Content
+	for _, m := range input.Messages {
+		if m.Role == ai.RoleSystem {
+			continue
+		}
+		parts, err := convertParts(m.Content)
+		if err != nil {
+			return nil, err
+		}
+		contents = append(contents, &genai.Content{
+			Parts: parts,
+			Role:  string(m.Role),
+		})
+	}
+
+	// Send out the actual request.
+	if cb == nil {
+		resp, err := client.Models.GenerateContent(ctx, model, contents, gc)
+		if err != nil {
+			return nil, err
+		}
+		r := translateResponse(resp)
+		r.Request = input
+		return r, nil
+	}
+
+	// Streaming version.
+	iter := client.Models.GenerateContentStream(ctx, model, contents, gc)
+	var r *ai.ModelResponse
+
+	// merge all streamed responses
+	var resp *genai.GenerateContentResponse
+	var chunks []string
+	for chunk, err := range iter {
+		// abort stream if error found in the iterator items
+		if err != nil {
+			return nil, err
+		}
+		for i, c := range chunk.Candidates {
+			tc := translateCandidate(c)
+			err := cb(ctx, &ai.ModelResponseChunk{
+				Content: tc.Message.Content,
+			})
+			if err != nil {
+				return nil, err
+			}
+			// stream only supports text
+			chunks = append(chunks, c.Content.Parts[i].Text)
+		}
+		// keep the last chunk for usage metadata
+		resp = chunk
+	}
+
+	// manually merge all candidate responses, iterator does not provide a
+	// merged response utility
+	merged := []*genai.Candidate{
+		{
+			Content: &genai.Content{
+				Parts: []*genai.Part{genai.NewPartFromText(strings.Join(chunks, ""))},
+			},
+		},
+	}
+	resp.Candidates = merged
+	r = translateResponse(resp)
+	if r == nil {
+		// No candidates were returned. Probably rare, but it might avoid a NPE
+		// to return an empty instead of nil result.
+		r = &ai.ModelResponse{}
+	}
+	r.Request = input
+	return r, nil
+}
+
+// convertRequest translates from [*ai.ModelRequest] to
+// *genai.GenerateContentParameters
+func convertRequest(client *genai.Client, model string, input *ai.ModelRequest) (*genai.GenerateContentConfig, error) {
+	gc := genai.GenerateContentConfig{}
+	gc.CandidateCount = genai.Ptr[int32](1)
+	if c, ok := input.Config.(*ai.GenerationCommonConfig); ok && c != nil {
```

Comment:
Need to make sure changes are properly merged from here: https://github.com/firebase/genkit/pull/2234

Depending on who merges first.